storage/replica_command.go

```diff
@@ -564,80 +585,80 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
 			args.Timestamp, consistent, args.Txn)
 
-		// Merge the results, the total ranges may be bigger than rangeCount.
+		// Merge the results. The total may be bigger than rangeCount, so we
+		// fix that.
 		kvs = append(kvs, revKvs...)
 		intents = append(intents, revIntents...)
+		if int64(len(kvs)) > rangeCount {
+			kvs = kvs[:rangeCount]
+		}
 	}
 	if err != nil {
 		// An error here is likely a WriteIntentError when reading consistently.
 		return reply, nil, err
 	}
-	if args.IgnoreIntents && len(intents) > 0 {
-		// NOTE (subtle): in general, we want to try to clean up dangling
-		// intents on meta records. However, if we're in the process of
-		// cleaning up a dangling intent on a meta record by pushing the
-		// transaction, we don't want to create an infinite loop:
-		//
-		// intent! -> push-txn -> range-lookup -> intent! -> etc...
-		//
-		// Instead we want:
-		//
-		// intent! -> push-txn -> range-lookup -> ignore intent, return old/new ranges
-		//
-		// On the range-lookup from a push transaction, we therefore
-		// want to suppress WriteIntentErrors and return a value
-		// anyway. But which value? We don't know whether the range
-		// update succeeded or failed, but if we don't return the
-		// correct range descriptor we may not be able to find the
-		// transaction to push. Since we cannot know the correct answer,
-		// we choose randomly between the pre- and post- transaction
-		// values. If we guess wrong, the client will try again and get
-		// the other value (within a few tries).
-		if rand.Intn(2) == 0 {
-			key, txn := intents[0].Key, &intents[0].Txn
-			val, _, err := engine.MVCCGet(batch, key, txn.Timestamp, true, txn)
-			if err != nil {
-				return reply, nil, err
-			}
+	// NOTE (subtle): in general, we want to try to clean up dangling
+	// intents on meta records. However, if we're in the process of
+	// cleaning up a dangling intent on a meta record by pushing the
+	// transaction, we don't want to create an infinite loop:
+	//
+	// intent! -> push-txn -> range-lookup -> intent! -> etc...
+	//
+	// Instead we want:
+	//
+	// intent! -> push-txn -> range-lookup -> ignore intent, return old/new ranges
+	//
+	// On the range-lookup from a push transaction, we therefore
+	// want to suppress WriteIntentErrors and return a value
+	// anyway. But which value? We don't know whether the range
+	// update succeeded or failed, but if we don't return the
+	// correct range descriptor we may not be able to find the
+	// transaction to push. Since we cannot know the correct answer,
+	// we choose randomly between the pre- and post- transaction
+	// values. If we guess wrong, the client will try again and get
+	// the other value (within a few tries).
+	//
+	// Additionally, if the inconsistent lookup returned no descriptor,
+	// we always return the intent if there is one.
+	//
+	// TODO(tschottdorf): Double-check that randomness is allowed here,
```

Comment:
I don't think this is incorrect. Read commands are only ever satisfied at the leader and they don't need replay protection, but the randomness is still bugging me a little.